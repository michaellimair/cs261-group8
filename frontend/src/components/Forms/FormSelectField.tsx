import {
  FormControl, FormErrorMessage, FormLabel, Select,
} from '@chakra-ui/react';
import { IFieldProps } from 'customTypes/form';
import { PropsWithChildren } from 'react';

interface IFormSelectFieldProps<T> extends Omit<IFieldProps<T>, 'type'> {
  options: {
    label: string;
    value: any;
  }[];
  placeholder?: string;
  label: string;
  valueAsNumber?: boolean;
}

const FormSelectField = <T extends any>({
  name,
  autoComplete,
  register,
  error,
  required,
  options,
  placeholder,
  label: formLabel,
  valueAsNumber,
}: PropsWithChildren<IFormSelectFieldProps<T>>) => (
  <FormControl
    id={name}
    data-testid={`${name}-formcontrol`}
    isRequired={required}
    isInvalid={Boolean(error)}
    mb={2}
  >
    <FormLabel data-testid={`${name}-label`}>{formLabel}</FormLabel>
    <Select
      placeholder={placeholder}
      {...register(name, {
        valueAsNumber,
      })}
      autoComplete={autoComplete}
    >
      {options.map(({ label, value }) => (
        <option value={value as any} key={value as any}>{label}</option>
      ))}
    </Select>
    <FormErrorMessage>
      {error}
    </FormErrorMessage>
  </FormControl>
  );

export default FormSelectField;
