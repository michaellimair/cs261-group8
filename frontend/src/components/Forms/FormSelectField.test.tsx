import { render, RenderResult } from '@testing-library/react';
import FormSelectField from './FormSelectField';

interface ISampleForm {
  field: string;
}

describe('components/FormSelectField', () => {
  const fieldName: keyof ISampleForm = 'field';
  const autoComplete = 'email';
  const label = 'customlabel';
  const placeholder = 'testplaceholder';
  let result: RenderResult<typeof import('@testing-library/dom/types/queries'), HTMLElement>;
  let registerFn: jest.Mock;

  beforeEach(() => {
    registerFn = jest.fn();
    result = render(
      <FormSelectField<ISampleForm>
        name={fieldName}
        autoComplete={autoComplete}
        register={registerFn}
        label={label}
        placeholder={placeholder}
        options={[
          {
            label: 'test',
            value: 'test',
          },
          {
            label: 'new',
            value: 'new',
          },
        ]}
      />,
    );
  });

  afterEach(() => {
    registerFn.mockReset();
    result.unmount();
  });

  it('renders correctly with the right set of attributes', async () => {
    const selectFormControl = result.queryByTestId(`${fieldName}-formcontrol`)!;
    expect(selectFormControl).not.toBeNull();
    expect(selectFormControl).toMatchSnapshot();

    const inputLabelField = selectFormControl.querySelector('label')!;
    expect(inputLabelField).toBeTruthy();
    expect(inputLabelField.innerHTML).toEqual(label);

    const selectField = selectFormControl.querySelector('select')!;
    expect(selectField).toBeTruthy();
    expect(selectField.id).toEqual(fieldName);
    expect(selectField.getAttribute('autoComplete')).toEqual(autoComplete);
    expect(selectField).toMatchSnapshot();
  });
});
