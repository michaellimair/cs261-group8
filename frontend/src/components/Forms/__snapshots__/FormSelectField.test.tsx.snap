// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/FormSelectField renders correctly with the right set of attributes 1`] = `
<div
  class="chakra-form-control css-wlea3r"
  data-testid="field-formcontrol"
  role="group"
>
  <label
    class="chakra-form__label css-uvho5g"
    data-testid="field-label"
    for="field"
    id="field-label"
  >
    customlabel
  </label>
  <div
    class="chakra-select__wrapper css-42b2qy"
  >
    <select
      autocomplete="email"
      class="chakra-select css-1ik61og"
      id="field"
    >
      <option
        value=""
      >
        testplaceholder
      </option>
      <option
        value="test"
      >
        test
      </option>
      <option
        value="new"
      >
        new
      </option>
    </select>
    <div
      class="chakra-select__icon-wrapper css-162mkon"
    >
      <svg
        aria-hidden="true"
        class="chakra-select__icon"
        focusable="false"
        role="presentation"
        style="width: 1em; height: 1em; color: currentColor;"
        viewBox="0 0 24 24"
      >
        <path
          d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`components/FormSelectField renders correctly with the right set of attributes 2`] = `
<select
  autocomplete="email"
  class="chakra-select css-1ik61og"
  id="field"
>
  <option
    value=""
  >
    testplaceholder
  </option>
  <option
    value="test"
  >
    test
  </option>
  <option
    value="new"
  >
    new
  </option>
</select>
`;
